core/src/services/s3/backend.rs

```diff
@@ -769,6 +769,11 @@ impl Builder for S3Builder {
         debug!("backend use region: {region}");
 
         // Building endpoint.
+        if let Some(ref v) = self.config.endpoint {
+            cfg.endpoint_url = Some(v.to_string());
+        } else if let Some(ref v) = cfg.endpoint_url {
```

Comment:
I suspect the `else if` here is not the intended behavior, instead, we need:

```rust
// Retain the user's endpoint if it exists; otherwise, try loading it from the environment.
self.config.endpoint = self.config.endpoint.or(cfg.endpoint_url);
```

Reply:
```rust
        if let Some(ref v) = self.config.endpoint {
            cfg.endpoint_url = Some(v.to_string());
        } else
```

The if-branch guarded `self.config.endpoint` is None in the else-branch.

Reply:
We don't need to sync `self.config.endpoint` back to `cfg.endpoint_url`.